storage/src/main/java/io/delta/storage/internal/S3LogStoreUtil.java

```diff
@@ -0,0 +1,119 @@
+        }
+
+        @Override
+        public String toString() {
+            return "ACCEPT_ALL";
+        }
+    };
+
+    /**
+     * Uses the S3ListRequest.v2 interface with the startAfter parameter to only list files
+     * which are lexicographically greater than resolvedPath.
+     */
+    public static RemoteIterator<S3AFileStatus> s3ListFrom(FileSystem fs, Path resolvedPath, Path parentPath) throws IOException {
+        S3AFileSystem s3afs = (S3AFileSystem) fs;
+        int maxKeys = S3AUtils.intOption(s3afs.getConf(), MAX_PAGING_KEYS, DEFAULT_MAX_PAGING_KEYS, 1);
+        Listing listing = s3afs.getListing();
+        // List files lexicographically after resolvedPath inclusive within the same directory
+        return listing.createFileStatusListingIterator(resolvedPath,
+                S3ListRequest.v2(
+                        new ListObjectsV2Request()
+                                .withBucketName(s3afs.getBucket())
+                                .withMaxKeys(maxKeys)
+                                .withPrefix(s3afs.pathToKey(parentPath))
+                                .withStartAfter(keyBefore(s3afs.pathToKey(resolvedPath)))
+                ), ACCEPT_ALL,
+                new Listing.AcceptAllButSelfAndS3nDirs(parentPath)
+        );
+    }
+
+    /**
+     * Uses the S3ListRequest.v2 interface with the startAfter parameter to only list files
+     * which are lexicographically greater than resolvedPath.
+     *
+     * Wraps s3ListFrom in an array. Contained in this class to avoid contaminating other
+     * classes with dependencies on recent Hadoop versions.
+     *
+     * TODO: Remove this method when iterators are used everywhere.
+     */
+    public static FileStatus[] s3ListFromArray(FileSystem fs, Path resolvedPath, Path parentPath) throws IOException {
+        try {
+            return iteratorToStatuses(S3LogStoreUtil.s3ListFrom(fs, resolvedPath, parentPath), new HashSet<>());
+        } catch (ClassCastException | NoClassDefFoundError e) {
+            throw new FastS3ListFromUnavailableException(e);
+        }
+    }
+
+    /**
+     * Get the key which is lexicographically right before key.
+     * If the key is empty return null.
+     * If the key ends in a null byte, remove the last byte.
+     * Otherwise, subtract one from the last byte.
+     */
+    static String keyBefore(String key) {
+        byte[] bytes = key.getBytes(StandardCharsets.UTF_8);
+        if(bytes.length == 0) return null;
+        if(bytes[bytes.length - 1] > 0) {
+            bytes[bytes.length - 1] -= 1;
+            return new String(bytes, StandardCharsets.UTF_8);
+        } else {
+            return new String(bytes, 0, bytes.length - 1, StandardCharsets.UTF_8);
+        }
+    }
+
+    /**
+     * Wrapper exception with more diagnostics if the fast s3 list from feature is enabled, but we are unable to use it.
+     */
+    public static class FastS3ListFromUnavailableException extends RuntimeException {
```

Comment:
nit: we don't need to introduce a new exception type. We can just use `UnsupportedOperationException`.